### ISSUE TITLE
Escaping/Parallel await for syntax

### ISSUE THREAD

**Issue by lukepighetti** [dart-lang/language#1440]
I'm not aware of any way to subscribe to two streams using `await for` in an `async*` block. There are stream combining tools made available in `rx_dart` but I believe there is a lot of benefit to making this available with the core async tools.

As it stands today, I am not aware of any way to combine these two streams in an `async*` block. 

```dart
Stream<BBQStatus> deviceStatus() async* {
  var status = BBQStatus.initial();

  /// Temperature probe events
  await for (var e in service.probeEvents()) {
    status = status.copyWith(probes: e);
    yield status;
  }

  /// Battery voltage events
  await for (var e in service.batteryEvents()) {
    status = status.copyWith(battery: e);
    yield status;
  }

  print('This is not reachable until both streams close');
}
```

You can combine streams using things `ZipStream` (but you lose types) or rx_dart's `combineLatestN`

`escaping` is not what I would suggest to use, but merely to illustrate the concept of being able to allow the program to progress beyond an `await for` subscription before it has completed.

```dart
Stream<BBQStatus> deviceStatus() async* {
  var status = BBQStatus.initial();

  /// Temperature probe events
  escaping await for (var e in service.probeEvents()) {
    status = status.copyWith(probes: e);
    yield status;
  }

  /// Battery voltage events
  escaping await for (var e in service.batteryEvents()) {
    status = status.copyWith(battery: e);
    yield status;
  }

  print('This is reachable immediately. This Stream will '
      'close when all escaping await for stream subscriptions complete'
      'or when any error');
}
```

**Comment by lukepighetti** [comment]
Adjusting the concept to match syntax proposed by https://github.com/dart-lang/language/issues/1441

```dart
Stream<BBQStatus> deviceStatus() async* {
  var status = BBQStatus.initial();

  /// Temperature probe events and battery voltage events
  await for (var e in service.probeEvents()) {
    status = status.copyWith(probes: e);
    yield status;
  } and (var e in service.batteryEvents()) {
    status = status.copyWith(battery: e);
    yield status;
  }

  print('This is reachable when  all escaping await for stream subscriptions complete '
      'or when any error is thrown');
}
```

**Comment by lukepighetti** [comment]
Combined with syntax proposed by https://github.com/dart-lang/language/issues/1441

```dart
Stream<BBQStatus> deviceStatus() async* {
  var status = BBQStatus.initial();

  /// Temperature probe events and battery voltage events
  await for (var e in service.probeEvents()) {
    status = status.copyWith(probes: e);
    yield status;
  } and (var e in service.batteryEvents()) {
    status = status.copyWith(battery: e);
    yield status;
  } on StreamException {
    ///
  } catch(e) {
    ///
  }
  

  print('This is reachable when probeEvents and batteryEvents subscriptions complete '
      'or any error block throws / breaks');
}
```

**Comment by lrhn** [comment]
What you're asking for is a *non blocking `await for` loop*.
You start the loop, but you *also* continue executing code after the loop.

The big question is: When do you *stop* looping? Can you break it from the outside, or does it keep running until the stream ends (or the code inside the loop breaks out).

What happens if your code inside the loop contains a `return`? Or a `break` of more than just that loop? It probably *can't* be allowed to do that, we can only have one flow of control at a time inside the same function invocation, so it's not a real loop, just a fancy way to write a callback function. (Fancy can be fine, it looks better in otherwise `async` code).

The code must keep running event if the main control flow reaches the end of the function. What if it reaches that by throwing? That would normally emit an error on the stream and close it, but now it can't close it any more. (That could be a way to have streams with embedded errors, if the loops can also emit errors and events later).

Basically, a way to do *something* in reaction to events, while staying inside the scope where you can still `yield` to the current stream. Could also work for futures, so you can await two futures in parallel.

It's not clear how to makes consistent semantics for something like this.

**Comment by lukepighetti** [comment]
When I first posted this issue it was fairly broad, but I honed it to a syntax which I believe handles the majority of the concerns raised.


> What you're asking for is a non blocking await for loop.
> You start the loop, but you also continue executing code after the loop.

The currently proposed syntax is still blocking, but it uses the `and` keyword to link two `await for` blocks together.

> The big question is: When do you stop looping? Can you break it from the outside, or does it keep running until the stream ends (or the code inside the loop breaks out).

The proposed syntax completes the `await for ... and ...` blocks when both blocks complete naturally or with `break`. Or it errors when either throw an error.

> What happens if your code inside the loop contains a return? Or a break of more than just that loop? It probably can't be allowed to do that, we can only have one flow of control at a time inside the same function invocation, so it's not a real loop, just a fancy way to write a callback function. (Fancy can be fine, it looks better in otherwise async code).

Can't return a value from a generator function (using the 'async*' modifier).

> The code must keep running event if the main control flow reaches the end of the function. What if it reaches that by throwing? That would normally emit an error on the stream and close it, but now it can't close it any more. (That could be a way to have streams with embedded errors, if the loops can also emit errors and events later).

The currently propose syntax is blocking.

> Basically, a way to do something in reaction to events, while staying inside the scope where you can still yield to the current stream. Could also work for futures, so you can await two futures in parallel.
> 
> It's not clear how to makes consistent semantics for something like this.

Please feel free to let me know if anything is still not handled.

**Comment by lrhn** [comment]
The change makes it two parallel loops.

I wouldn't just allow it in a generator function, it should work in plain `async` functions too. If it's useful, it's probably useful in more ways than one. It only works for async loops (`await for`), so at least it's `async` or `async*` only.

There are multiple ways to exit a loop:
* complete normally,
* throw,
* break the loop,
* break an outer statement,
* continue an outer loop/switch, or
* return, with or without a value. (Also happens on a `yield` in a cancelled stream).

Breaking the loop itself or completing normally are fine cases to just wait for the other loop.

In all the other cases, it should probably break both loops when control flow exits the combined `await for (  ) { } and ( ) { }` statement.

If you combine it with #1441 (and even #171), I'd probably want individual exception handlers on each loop, so:
```dart
await for (var e in stream) {
  doSomething(e);
} catch (e) {
  doSomethingElse(e);
} else {
  doSomethingAtTheEnd();
} 
and for (var e in otherStream) {
  doWhatNot(e);
} catch (e) {
  doWhatNotElse(e);  
} else {
  doWhatNotAtTheEnd();
}
```
so the grammar would be:
```ebnf
<asyncLoop> ::= `await` <forInLoop> (`and` <forInLoop>)*
```
where it's a compile-time error if the expression of the `for`/`in` is not assignable to a `Stream` type.

The `catch`/`on` clauses apply to all `for`/`in` loops (it's less common to catch errors from `get current` in non-async loops than to get error events from streams, but there is no need to prevent it).

**Comment by lukepighetti** [comment]
LGTM
